src/apps/external_executor/external_executor_indexing.cpp

```diff
@@ -0,0 +1,194 @@
+              return std::nullopt;
+              break;
+            }
+          }
+        }
+      }
+      else
+      {
+        // We're not currently fetching this. First check if it's in our
+        // current results
+        const auto current_it = indexed_data.find(key);
+        if (current_it.has_value())
+        {
+          if (complete)
+          {
+            return current_it.value();
+          }
+        }
+        else
+        {
+          // Begin fetching this bucket from disk
+          std::string blob_name = get_blob_name(map_name, key);
+          auto fetch_handle = lfs_access->fetch(blob_name);
+          std::string value;
+          results_in_progress[key] = std::make_pair(fetch_handle, value);
+          complete = false;
+        }
+      }
+    }
+  }
+
+  void MapIndex::store_data(std::string& key, std::string& value)
+  {
+    indexed_data.insert(key, std::move(value));
+  }
+
+  ExecutorIndex::ExecutorIndex(
+    const std::string& map_name_,
+    const std::string& strategy_prefix,
+    IndexDataStructure ds,
+    ExecutorId& id,
+    ccf::endpoints::CommandEndpointContext& ctx,
+    ccfapp::AbstractNodeContext& node_ctx,
+    IndexStream&& stream) :
+    Strategy(strategy_prefix),
+    map_name(map_name_),
+    data_structure(ds),
+    indexer_id(id),
+    endpoint_ctx(&ctx),
+    node_context(&node_ctx),
+    out_stream(std::move(stream)),
+    is_indexer_active(true)
+  {
+    if (kv::get_security_domain(map_name_) != kv::SecurityDomain::PUBLIC)
+    {
+      throw std::logic_error(fmt::format(
+        "This Strategy ({}) is currently only implemented for public tables, "
+        "so cannot be used for '{}'",
+        get_name(),
+        map_name_));
+    }
+    if (data_structure == MAP)
+    {
+      impl_index = std::make_unique<MapIndex>(
+        node_context->get_subsystem<ccf::indexing::AbstractLFSAccess>(),
+        map_name);
+    }
+    else if (data_structure == PREFIX_TREE)
+    {
+      impl_index = std::make_unique<PrefixTreeIndex>();
+    }
+
+    // create a detached stream pointer of the indexer
+    detached_stream =
+      ccf::grpc::detach_stream(ctx.rpc_ctx, std::move(out_stream), [this]() {
+        is_indexer_active = false;
+      });
+  }
+
+  void ExecutorIndex::handle_committed_transaction(
+    const ccf::TxID& tx_id, const kv::ReadOnlyStorePtr& store)
+  {
+    auto tx = store->create_read_only_tx();
+    auto handle = tx.ro<Map>(map_name);
+
+    handle->foreach([this](const auto& k, const auto& v) {
+      externalexecutor::protobuf::IndexWork data;
+      externalexecutor::protobuf::IndexKeyValue* index_key_value =
+        data.mutable_key_value();
+      index_key_value->set_key(k);
+      index_key_value->set_value(v);
+      if (is_indexer_active)
+      {
+        // stream transactions to the indexer
+        if (!detached_stream->stream_msg(data))
+        {
+          LOG_DEBUG_FMT("Failed to stream request to indexer {}", indexer_id);
+        }
```

Comment:
I think this should be `LOG_FAIL`, and we should early-out here - any failure probably means every future message will also fail, and even if it didn't we don't want to continue with missing entries!

Reply:
Good point. Failing early is better here.